allel/stats/roh.py

```diff
@@ -0,0 +1,231 @@
+        from allel.stats import windowed_statistic
+
+        window = np.arange(0, is_accessible.size, plotwinsize)
+        windows = np.vstack([window[:-1], window[1:]]).T
+
+        count_hz, win, count = windowed_statistic(pos=pos, values=hz,
+                                                  statistic=np.sum,
+                                                  windows=windows)
+
+        if is_accessible is not None:
+
+            # remove the last val as not a complete window
+            count_acc = np.add.reduceat(
+                is_accessible, window)[:-1]
+
+            # if accessible bases in a window < 1000, don't plot.
+            # b/c Leads to spiky data
+            count_acc[count_acc < 1000] = 0
+            count_hz[count_acc < 1000] = 0
+            heterozyg = count_hz/count_acc
+
+            assert not np.any(count_hz > count_acc), \
+                "heterozygosity exceeds 1.0. Check accessibility array input"
+        else:
+            heterozyg = count_hz/plotwinsize
+
+        x = win.mean(axis=1) * 1e-6
+
+        plt.figure(figsize=(6, 1.5))
+        plt.plot(x, heterozyg, "b-", label="Heterozygosity")
+        plt.ylabel('Heterozygosity')
+
+        for low, upp in roh * 1e-6:
+            plt.axvspan(low, upp, facecolor='g', alpha=0.5, edgecolor='none')
+
+        if is_accessible is not None:
+            inaccessible_windows = _get_state_windows(is_accessible)
+
+            for low, upp in (inaccessible_windows * 1e-6):
+                # don't plot inaccessible gaps < 5kb
+                if (upp - low) > .005:
+                    plt.axvspan(low, upp, facecolor='m', alpha=0.1,
+                                edgecolor='none')
+
+        plt.xlabel('Position (Mbp)')
+        plt.title(" ".join(['Called runs of homozygosity.',
+                            r'$F_{ROH}:$',
+                            "{0:.4f}".format(froh)]))
+        plt.tick_params(top="off", right="off")
+        plt.legend()
+
+    # return roh and froh
+    return roh, froh
+
+
+def _predict_roh_state(model, is_het, pos, is_accessible=None,
+                       contig_size=None):
+
+    # assume non-included pos are homozygous reference
+    assert is_het.shape == pos.shape
+
+    # declaration of this object may result in large mem usage
+    if is_accessible is not None:
+        contig_size = is_accessible.size
+
+    observations = np.zeros((contig_size, 1), dtype='int')
+
+    # these are hets
+    observations[np.compress(is_het, pos) - 1] = 1
+
+    # these are unobserved
+    if is_accessible is not None:
+        observations[~is_accessible] = 2
+
+    predictions = model.predict(X=observations)
+    return predictions, observations
+
+
+def _derive_emission_mx(prob, pa):
+    # one row per p in prob
+    # hom, het, unobserved
+    mx = [[(1-p)*pa, p*pa, 1-pa] for p in prob]
+    mxe = np.array(mx)
+    assert mxe.shape == (prob.size, 3)
+    return mxe
+
+
+def _derive_transition_mx(pr, nstates):
```

Comment:
Suggest to rename _mhmm_derive_transition_mx to leave room for other roh-calling functions in future.